File: apps/ios/PeerScholar/Views/Components/Components.swift
import SwiftUI

struct CardBackground: ViewModifier {
    func body(content: Content) -> some View {
        content
            .padding(14)
            .background(Color(.secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
    }
}

extension View {
    func cardBackground() -> some View { modifier(CardBackground()) }
}

// Deterministic stand-in photography — same source picsum.photos/pravatar.cc
// used by the website, so every front end demos with the same imagery.
enum RemoteImage {
    static func photo(seed: String, width: Int = 480, height: Int = 320) -> URL? {
        URL(string: "https://picsum.photos/seed/\(seed)/\(width)/\(height)")
    }
    static func avatar(seed: String, size: Int = 128) -> URL? {
        URL(string: "https://i.pravatar.cc/\(size)?u=\(seed)")
    }
}

struct ScaleButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.97 : 1)
            .animation(.spring(response: 0.25, dampingFraction: 0.6), value: configuration.isPressed)
    }
}

struct RatingStars: View {
    let rating: Double
    var count: Int? = nil

    var body: some View {
        if rating == 0 {
            Text("No ratings yet").font(.caption).foregroundStyle(.secondary)
        } else {
            HStack(spacing: 4) {
                Image(systemName: "star.fill").foregroundStyle(.yellow).font(.caption)
                Text(String(format: "%.1f", rating)).font(.subheadline.weight(.semibold))
                if let count {
                    Text("(\(count))").font(.caption).foregroundStyle(.secondary)
                }
            }
        }
    }
}

struct SubjectTag: View {
    let subject: String
    var body: some View {
        Text(subject.uppercased())
            .font(.caption2.weight(.bold))
            .foregroundStyle(Color.brand)
    }
}

struct Thumbnail: View {
    let seed: String
    var height: CGFloat = 110
    var badge: String? = nil

    var body: some View {
        ZStack(alignment: .bottomLeading) {
            AsyncImage(url: RemoteImage.photo(seed: seed)) { phase in
                switch phase {
                case .success(let image):
                    image.resizable().scaledToFill()
                default:
                    LinearGradient(colors: [Color.brandNavy, Color.brand], startPoint: .topLeading, endPoint: .bottomTrailing)
                }
            }
            .frame(height: height)
            .frame(maxWidth: .infinity)
            .clipped()

            LinearGradient(colors: [.black.opacity(0.65), .clear], startPoint: .bottom, endPoint: .center)
                .frame(height: height)

            if let badge {
                Text(badge.uppercased())
                    .font(.caption2.weight(.bold))
                    .foregroundStyle(.white)
                    .padding(.horizontal, 10)
                    .padding(.bottom, 8)
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
    }
}

struct AvatarImage: View {
    let seed: String
    var size: CGFloat = 48

    var body: some View {
        AsyncImage(url: RemoteImage.avatar(seed: seed, size: Int(size * 2))) { phase in
            switch phase {
            case .success(let image):
                image.resizable().scaledToFill()
            default:
                Circle().fill(LinearGradient(colors: [Color.brandNavy, Color.brand], startPoint: .top, endPoint: .bottom))
            }
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay(Circle().stroke(.white, lineWidth: 2))
        .shadow(color: .black.opacity(0.1), radius: 3, y: 1)
    }
}

struct CourseRow: View {
    let course: Course
    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Thumbnail(seed: course.id, badge: course.subject)
            VStack(alignment: .leading, spacing: 4) {
                Text(course.title).font(.headline)
                Text("by \(course.tutorName)").font(.subheadline).foregroundStyle(.secondary)
                HStack {
                    RatingStars(rating: course.ratingAvg, count: course.ratingCount)
                    Spacer()
                    Text(course.priceDisplay).font(.subheadline.weight(.bold))
                }
            }
            .padding(12)
        }
        .background(Color(.secondarySystemGroupedBackground))
        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
    }
}

struct TutorRow: View {
    let tutor: Profile
    var body: some View {
        HStack(spacing: 12) {
            AvatarImage(seed: tutor.id, size: 48)
            VStack(alignment: .leading, spacing: 4) {
                Text(tutor.fullName).font(.headline)
                RatingStars(rating: tutor.ratingAvg, count: tutor.ratingCount)
                Text(tutor.bio).font(.caption).foregroundStyle(.secondary).lineLimit(2)
            }
        }
        .cardBackground()
    }
}

struct SessionCard: View {
    let session: LiveSession
    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Thumbnail(seed: session.id, height: 90, badge: "● Live")
            VStack(alignment: .leading, spacing: 6) {
                SubjectTag(subject: session.subject)
                Text(session.title).font(.headline)
                Text("with \(session.tutorName)").font(.subheadline).foregroundStyle(.secondary)
                Text(session.scheduledAt.formatted(date: .abbreviated, time: .shortened) + " · \(session.durationMinutes) min")
                    .font(.caption).foregroundStyle(.secondary)
                HStack {
                    Text("\(session.spotsLeft) of \(session.maxParticipants) spots left")
                        .font(.caption).foregroundStyle(.secondary)
                    Spacer()
                    Text(session.priceDisplay).font(.subheadline.weight(.bold))
                }
                Button("Book session") {}
                    .buttonStyle(ScaleButtonStyle())
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 10)
                    .background(Color.brand)
                    .foregroundStyle(.white)
                    .fontWeight(.semibold)
                    .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
            }
            .padding(12)
        }
        .background(Color(.secondarySystemGroupedBackground))
        .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
    }
}

/// Fades and slides content up on first appearance — used to give scroll
/// views a gentler, staggered entrance instead of popping in instantly.
/// Skips the motion entirely when the person has Reduce Motion turned on
/// (Design Guideline — Accessibility > Cognitive).
struct AppearAnimation: ViewModifier {
    var delay: Double = 0
    @State private var shown = false
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func body(content: Content) -> some View {
        content
            .opacity(reduceMotion ? 1 : (shown ? 1 : 0))
            .offset(y: reduceMotion ? 0 : (shown ? 0 : 12))
            .onAppear {
                guard !reduceMotion else { return }
                withAnimation(.easeOut(duration: 0.5).delay(delay)) {
                    shown = true
                }
            }
    }
}

extension View {
    func appearAnimation(delay: Double = 0) -> some View { modifier(AppearAnimation(delay: delay)) }
}

File: apps/android/app/src/main/java/org/peerscholar/app/ui/components/Components.kt
package org.peerscholar.app.ui.components

import android.content.Intent
import android.net.Uri
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.graphicsLayer
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import org.peerscholar.app.data.AppStore
import org.peerscholar.app.model.Course
import org.peerscholar.app.model.LiveSession
import org.peerscholar.app.model.Profile
import org.peerscholar.app.ui.theme.Brand600

// Deterministic stand-in photography — same source (picsum.photos /
// i.pravatar.cc) used by the website and iOS app, so every front end demos
// with the same imagery.
private fun photoUrl(seed: String, w: Int = 480, h: Int = 320) = "https://picsum.photos/seed/$seed/$w/$h"
private fun avatarUrl(seed: String, size: Int = 128) = "https://i.pravatar.cc/$size?u=$seed"

@Composable
fun Card(modifier: Modifier = Modifier, content: @Composable ColumnScope.() -> Unit) {
    ElevatedCard(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(14.dp),
        colors = CardDefaults.elevatedCardColors(containerColor = MaterialTheme.colorScheme.surface),
    ) {
        Column(Modifier.padding(14.dp), content = content)
    }
}

@Composable
fun Thumbnail(seed: String, modifier: Modifier = Modifier, height: Int = 110, badge: String? = null) {
    Box(modifier.fillMaxWidth().height(height.dp).clip(RoundedCornerShape(14.dp))) {
        AsyncImage(
            model = photoUrl(seed),
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier.fillMaxSize(),
        )
        Box(
            Modifier
                .fillMaxSize()
                .background(Brush.verticalGradient(listOf(Color.Transparent, Color.Black.copy(alpha = 0.6f))))
        )
        if (badge != null) {
            Text(
                badge.uppercase(),
                color = Color.White,
                fontWeight = FontWeight.Bold,
                style = MaterialTheme.typography.labelSmall,
                modifier = Modifier.align(Alignment.BottomStart).padding(10.dp),
            )
        }
    }
}

@Composable
fun Avatar(seed: String, size: Int = 48, modifier: Modifier = Modifier) {
    AsyncImage(
        model = avatarUrl(seed, size * 2),
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = modifier.size(size.dp).clip(CircleShape),
    )
}

/** Fades and slides a row in on first composition, staggered by [index]. */
@Composable
fun Modifier.appearAnimation(index: Int = 0): Modifier {
    var shown by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) {
        kotlinx.coroutines.delay(index * 60L)
        shown = true
    }
    val alpha by animateFloatAsState(if (shown) 1f else 0f, animationSpec = tween(400), label = "appearAlpha")
    val offsetY by animateFloatAsState(if (shown) 0f else 24f, animationSpec = tween(400), label = "appearOffset")
    return this
        .graphicsLayer(alpha = alpha, translationY = offsetY)
}

@Composable
fun RatingStars(rating: Double, count: Int? = null) {
    if (rating == 0.0) {
        Text("No ratings yet", style = MaterialTheme.typography.bodySmall, color = Color.Gray)
        return
    }
    Row(verticalAlignment = Alignment.CenterVertically) {
        Icon(Icons.Filled.Star, contentDescription = null, tint = Color(0xFFF59E0B), modifier = Modifier.size(14.dp))
        Spacer(Modifier.width(4.dp))
        Text("%.1f".format(rating), fontWeight = FontWeight.SemiBold, style = MaterialTheme.typography.bodySmall)
        count?.let {
            Spacer(Modifier.width(4.dp))
            Text("($it)", style = MaterialTheme.typography.bodySmall, color = Color.Gray)
        }
    }
}

@Composable
fun SubjectTag(subject: String) {
    Text(subject.uppercase(), color = Brand600, fontWeight = FontWeight.Bold, style = MaterialTheme.typography.labelSmall)
}

@Composable
fun CourseRow(course: Course, onClick: () -> Unit) {
    ElevatedCard(
        modifier = Modifier.fillMaxWidth().clickable(onClick = onClick),
        shape = RoundedCornerShape(14.dp),
        colors = CardDefaults.elevatedCardColors(containerColor = MaterialTheme.colorScheme.surface),
    ) {
        Column {
            Thumbnail(course.id, badge = course.subject)
            Column(Modifier.padding(14.dp)) {
                Text(course.title, style = MaterialTheme.typography.titleMedium)
                Text("by ${course.tutorName}", style = MaterialTheme.typography.bodySmall, color = Color.Gray)
                Spacer(Modifier.height(6.dp))
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween, verticalAlignment = Alignment.CenterVertically) {
                    RatingStars(course.ratingAvg, course.ratingCount)
                    Text(course.priceDisplay, fontWeight = FontWeight.Bold)
                }
            }
        }
    }
}

@Composable
fun TutorRow(tutor: Profile, onClick: () -> Unit) {
    Card(Modifier.clickable(onClick = onClick)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Avatar(tutor.id, size = 48)
            Spacer(Modifier.width(12.dp))
            Column {
                Text(tutor.fullName, style = MaterialTheme.typography.titleMedium)
                RatingStars(tutor.ratingAvg, tutor.ratingCount)
                Text(tutor.bio, style = MaterialTheme.typography.bodySmall, maxLines = 2, color = Color.Gray)
            }
        }
    }
}

@Composable
fun SessionCard(session: LiveSession, store: AppStore) {
    // Note: collected unconditionally — calling collectAsState() through a
    // null-safe chain would make a composable invocation conditional, which
    // breaks Compose's positional memoization.
    val state by store.state.collectAsState()
    val booked = state.bookings.contains(session.id)
    val spotsLeft = (session.spotsLeft - if (booked) 1 else 0).coerceAtLeast(0)
    val isFull = spotsLeft <= 0 && !booked
    val isFree = session.priceCents == 0
    val context = LocalContext.current
    var showConfirm by remember { mutableStateOf(false) }

    ElevatedCard(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(14.dp),
        colors = CardDefaults.elevatedCardColors(containerColor = MaterialTheme.colorScheme.surface),
    ) {
        Column {
            Box {
                Thumbnail(session.id, height = 90, badge = "● Live")
                if (booked) {
                    Text(
                        "BOOKED",
                        color = Color.White,
                        fontWeight = FontWeight.Bold,
                        style = MaterialTheme.typography.labelSmall,
                        modifier = Modifier
                            .align(Alignment.TopEnd)
                            .padding(8.dp)
                            .background(Color(0xFF16A34A), RoundedCornerShape(20.dp))
                            .padding(horizontal = 8.dp, vertical = 3.dp),
                    )
                }
            }
            Column(Modifier.padding(14.dp)) {
                SubjectTag(session.subject)
                Text(session.title, style = MaterialTheme.typography.titleMedium)
                Text("with ${session.tutorName}", style = MaterialTheme.typography.bodySmall, color = Color.Gray)
                Text(
                    "${session.durationMinutes} min",
                    style = MaterialTheme.typography.bodySmall, color = Color.Gray
                )
                Spacer(Modifier.height(6.dp))
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                    Text(
                        if (isFull) "Fully booked" else "$spotsLeft of ${session.maxParticipants} spots left",
                        style = MaterialTheme.typography.bodySmall,
                        color = if (isFull) Color(0xFFE11D48) else Color.Gray,
                    )
                    Text(if (isFree) "Free" else session.priceDisplay, fontWeight = FontWeight.Bold)
                }
                Spacer(Modifier.height(8.dp))

                if (booked) {
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        session.joinUrl?.let { url ->
                            Button(
                                onClick = {
                                    context.startActivity(
                                        Intent(Intent.ACTION_VIEW, Uri.parse(url))
                                    )
                                },
                                modifier = Modifier.weight(1f),
                                colors = ButtonDefaults.buttonColors(containerColor = Brand600),
                            ) { Text("Join") }
                        }
                        OutlinedButton(onClick = { store.cancelBooking(session.id) }) { Text("Cancel") }
                    }
                } else {
                    Button(
                        onClick = { showConfirm = true },
                        enabled = !isFull,
                        modifier = Modifier.fillMaxWidth(),
                        colors = ButtonDefaults.buttonColors(containerColor = Brand600),
                    ) {
                        Text(if (isFull) "Fully booked" else "Book session")
                    }
                }
            }
        }
    }

    if (showConfirm) {
        AlertDialog(
            onDismissRequest = { showConfirm = false },
            title = { Text("Confirm your booking") },
            text = {
                Column {
                    Text(session.title, fontWeight = FontWeight.SemiBold)
                    Text(
                        "with ${session.tutorName} · ${session.durationMinutes} minutes",
                        style = MaterialTheme.typography.bodySmall,
                        color = Color.Gray,
                    )
                    Spacer(Modifier.height(12.dp))
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Text("Total", fontWeight = FontWeight.Bold)
                        Text(if (isFree) "$0.00" else session.priceDisplay, fontWeight = FontWeight.Bold)
                    }
                    if (!isFree) {
                        Spacer(Modifier.height(10.dp))
                        Text(
                            "This prototype doesn't process real payments.",
                            style = MaterialTheme.typography.labelSmall,
                            color = Color.Gray,
                        )
                    }
                }
            },
            confirmButton = {
                Button(
                    onClick = {
                        store.book(session.id)
                        showConfirm = false
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = Brand600),
                ) { Text("Confirm booking") }
            },
            dismissButton = { TextButton(onClick = { showConfirm = false }) { Text("Cancel") } },
        )
    }
}
